File: src/main/java/net/arvandor/numinoustreasury/item/NuminousItemStack.java
package net.arvandor.numinoustreasury.item;

import static net.md_5.bungee.api.ChatColor.GRAY;
import static org.bukkit.persistence.PersistentDataType.STRING;
import static org.bukkit.persistence.PersistentDataType.TAG_CONTAINER_ARRAY;

import com.rpkit.core.service.Services;
import net.arvandor.numinoustreasury.NuminousTreasury;
import net.arvandor.numinoustreasury.item.log.NuminousLogEntry;
import org.bukkit.Bukkit;
import org.bukkit.configuration.serialization.ConfigurationSerializable;
import org.bukkit.configuration.serialization.SerializableAs;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.ItemMeta;
import org.bukkit.persistence.PersistentDataContainer;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Map;

@SerializableAs("NuminousItemStack")
public final class NuminousItemStack implements ConfigurationSerializable {

    private final NuminousItemType itemType;
    private final int amount;
    private final ItemStack[] inventoryContents;
    private final List<NuminousLogEntry> logEntries;

    public NuminousItemStack(NuminousItemType itemType, int amount, ItemStack[] inventoryContents, List<NuminousLogEntry> logEntries) {
        this.itemType = itemType;
        this.amount = amount;
        this.inventoryContents = inventoryContents;
        this.logEntries = logEntries;
    }

    public NuminousItemStack(NuminousItemType itemType, int amount) {
        this(itemType, amount, null, List.of());
    }

    public NuminousItemStack(NuminousItemType itemType) {
        this(itemType, 1);
    }

    public NuminousItemType getItemType() {
        return itemType;
    }

    public int getAmount() {
        return amount;
    }

    public ItemStack[] getInventoryContents() {
        return inventoryContents;
    }

    public List<NuminousLogEntry> getLogEntries() {
        return logEntries;
    }

    public ItemStack toItemStack() {
        return itemType.toItemStack(amount, logEntries);
    }

    public static NuminousItemStack fromItemStack(ItemStack itemStack) {
        if (itemStack == null) return null;
        NuminousTreasury plugin = (NuminousTreasury) Bukkit.getPluginManager().getPlugin("numinous-treasury");
        ItemMeta meta = itemStack.getItemMeta();
        if (meta == null) {
            return null;
        }
        String itemId = meta.getPersistentDataContainer().get(plugin.keys().itemId(), STRING);
        if (itemId == null) return null;
        NuminousItemService itemService = Services.INSTANCE.get(NuminousItemService.class);
        NuminousItemType itemType = itemService.getItemTypeById(itemId);
        if (itemType == null) return null;
        int amount = itemStack.getAmount();

        ItemStack[] inventoryContents = null;
        PersistentDataContainer[] itemTagContainers = meta.getPersistentDataContainer().get(plugin.keys().inventory(), TAG_CONTAINER_ARRAY);
        if (itemTagContainers != null) {
            inventoryContents = Arrays.stream(itemTagContainers)
                    .map(itemTagContainer -> itemTagContainer.get(
                            plugin.keys().inventoryItem(),
                            plugin.persistentDataTypes().itemStack()
                    )).toArray(ItemStack[]::new);
        }

        List<NuminousLogEntry> logEntries = meta.getPersistentDataContainer().has(plugin.keys().logEntries(), TAG_CONTAINER_ARRAY)
                ? Arrays.stream(meta.getPersistentDataContainer().get(plugin.keys().logEntries(), TAG_CONTAINER_ARRAY))
                .map(compoundTag -> NuminousLogEntry.fromCompoundTag(plugin, compoundTag))
                .toList()
                : List.of();
        return new NuminousItemStack(itemType, amount, inventoryContents, logEntries);
    }

    public void update(ItemStack itemStack, boolean updateLore) {
        if (itemStack == null) return;
        NuminousTreasury plugin = (NuminousTreasury) Bukkit.getPluginManager().getPlugin("numinous-treasury");

        itemStack.setType(itemType.getMinecraftItem().getType());
        itemStack.setAmount(amount);

        ItemMeta meta = itemStack.getItemMeta();
        if (meta == null) {
            return;
        }
        meta.getPersistentDataContainer().set(plugin.keys().itemId(), STRING, itemType.getId());
        meta.getPersistentDataContainer().set(plugin.keys().inventory(), TAG_CONTAINER_ARRAY, Arrays.stream(inventoryContents)
                .map(item -> {
                    PersistentDataContainer container = meta.getPersistentDataContainer().getAdapterContext().newPersistentDataContainer();
                    if (item != null) {
                        container.set(plugin.keys().inventoryItem(), plugin.persistentDataTypes().itemStack(), item);
                    }
                    return container;
                }).toArray(PersistentDataContainer[]::new));
        meta.getPersistentDataContainer().set(plugin.keys().logEntries(), TAG_CONTAINER_ARRAY, logEntries.stream()
                .map(logEntry -> logEntry.toCompoundTag(plugin, meta.getPersistentDataContainer()))
                .toArray(PersistentDataContainer[]::new));

        if (updateLore) {
            ItemMeta typeMeta = itemType.getMinecraftItem().getItemMeta();
            if (typeMeta != null) {
                meta.setDisplayName(typeMeta.getDisplayName());
            }

            List<String> lore = new ArrayList<>();
            NuminousRarity rarity = itemType.getRarity();
            lore.add(0, rarity.getColor() + rarity.getDisplayName());
            lore.add(1, GRAY + "Weight: " + itemType.getWeight().toString());
            meta.setLore(lore);
        }

        itemStack.setItemMeta(meta);
    }

    public NuminousItemStack copy(
            NuminousItemType itemType,
            Integer amount,
            ItemStack[] inventoryContents,
            List<NuminousLogEntry> logEntries
    ) {
        return new NuminousItemStack(
                itemType == null ? this.itemType : itemType,
                amount == null ? this.amount : amount,
                inventoryContents == null ? this.inventoryContents : inventoryContents,
                logEntries == null ? this.logEntries : logEntries
        );
    }

    @Override
    public Map<String, Object> serialize() {
        return Map.of(
                "item-type", getItemType().getId(),
                "amount", getAmount(),
                "inventory-contents", getInventoryContents(),
                "log-entries", getLogEntries()
        );
    }

    public static NuminousItemStack deserialize(Map<String, Object> serialized) {
        NuminousItemService itemService = Services.INSTANCE.get(NuminousItemService.class);
        List<ItemStack> inventory = (List<ItemStack>) serialized.get("inventory-contents");
        return new NuminousItemStack(
                itemService.getItemTypeById((String) serialized.get("item-type")),
                (Integer) serialized.get("amount"),
                inventory == null ? null : inventory.toArray(ItemStack[]::new),
                (List<NuminousLogEntry>) serialized.get("log-entries")
        );
    }
}

File: src/main/kotlin/net/arvandor/numinoustreasury/item/NuminousItemStack.kt
package net.arvandor.numinoustreasury.item

import com.rpkit.core.service.Services
import net.arvandor.numinoustreasury.NuminousTreasury
import net.arvandor.numinoustreasury.item.log.NuminousLogEntry
import net.md_5.bungee.api.ChatColor
import org.bukkit.Bukkit
import org.bukkit.configuration.serialization.ConfigurationSerializable
import org.bukkit.configuration.serialization.SerializableAs
import org.bukkit.inventory.ItemStack
import org.bukkit.persistence.PersistentDataContainer
import org.bukkit.persistence.PersistentDataType
import org.jetbrains.annotations.Contract

@SerializableAs("NuminousItemStack")
class NuminousItemStack
    @JvmOverloads
    constructor(
        val itemType: NuminousItemType,
        val amount: Int = 1,
        val inventoryContents: Array<ItemStack?>? = null,
        logEntries: List<NuminousLogEntry>? = listOf(),
    ) : ConfigurationSerializable {
        val logEntries: List<NuminousLogEntry>? = if (itemType.isAllowLogEntries) logEntries else null

        fun toItemStack(): ItemStack {
            return itemType.toItemStack(amount, logEntries)
        }

        fun update(
            itemStack: ItemStack?,
            updateLore: Boolean,
        ) {
            if (itemStack == null) return
            val plugin = Bukkit.getPluginManager().getPlugin("numinous-treasury") as NuminousTreasury

            itemStack.type = itemType.minecraftItem.type
            itemStack.amount = amount

            val meta = itemStack.itemMeta ?: return
            meta.persistentDataContainer.set(plugin.keys.itemId, PersistentDataType.STRING, itemType.id)
            meta.persistentDataContainer.set<Array<PersistentDataContainer>, Array<PersistentDataContainer>>(
                plugin.keys.inventory,
                PersistentDataType.TAG_CONTAINER_ARRAY,
                inventoryContents
                    ?.map { item: ItemStack? ->
                        val container = meta.persistentDataContainer.adapterContext.newPersistentDataContainer()
                        if (item != null) {
                            container.set(
                                plugin.keys.inventoryItem,
                                plugin.persistentDataTypes.itemStack,
                                item,
                            )
                        }
                        container
                    }
                    ?.toTypedArray()
                    ?: arrayOf(),
            )
            if (logEntries != null) {
                meta.persistentDataContainer.set<Array<PersistentDataContainer>, Array<PersistentDataContainer>>(
                    plugin.keys.logEntries,
                    PersistentDataType.TAG_CONTAINER_ARRAY,
                    logEntries
                        .map { logEntry ->
                            logEntry.toCompoundTag(
                                plugin,
                                meta.persistentDataContainer,
                            )
                        }
                        .toTypedArray(),
                )
            }

            if (updateLore) {
                val typeMeta = itemType.minecraftItem.itemMeta
                if (typeMeta != null) {
                    meta.setDisplayName(typeMeta.displayName)
                }

                val rarity = itemType.rarity
                meta.lore =
                    listOf(
                        rarity.color.toString() + rarity.displayName,
                        ChatColor.GRAY.toString() + "Weight: " + itemType.weight.toString(),
                    )
            }

            itemStack.setItemMeta(meta)
        }

        fun copy(
            itemType: NuminousItemType?,
            amount: Int?,
            inventoryContents: Array<ItemStack?>?,
            logEntries: List<NuminousLogEntry>?,
        ): NuminousItemStack {
            return NuminousItemStack(
                itemType ?: this.itemType,
                amount ?: this.amount,
                inventoryContents ?: this.inventoryContents,
                logEntries ?: this.logEntries,
            )
        }

        override fun serialize(): Map<String, Any?> {
            return mapOf(
                "item-type" to itemType.id,
                "amount" to amount,
                "inventory-contents" to inventoryContents,
                "log-entries" to logEntries,
            )
        }

        companion object {
            @Contract("null -> null")
            fun fromItemStack(itemStack: ItemStack?): NuminousItemStack? {
                if (itemStack == null) return null
                val plugin = Bukkit.getPluginManager().getPlugin("numinous-treasury") as NuminousTreasury
                val meta = itemStack.itemMeta ?: return null
                val itemId =
                    meta.persistentDataContainer.get(
                        plugin.keys.itemId,
                        PersistentDataType.STRING,
                    )
                if (itemId == null) return null
                val itemService =
                    Services.INSTANCE.get(
                        NuminousItemService::class.java,
                    )
                val itemType = itemService.getItemTypeById(itemId) ?: return null
                val amount = itemStack.amount

                var inventoryContents: Array<ItemStack?>? = null
                val itemTagContainers =
                    meta.persistentDataContainer.get(
                        plugin.keys.inventory,
                        PersistentDataType.TAG_CONTAINER_ARRAY,
                    )
                if (itemTagContainers != null) {
                    inventoryContents =
                        itemTagContainers
                            .map { itemTagContainer ->
                                itemTagContainer.get(
                                    plugin.keys.inventoryItem,
                                    plugin.persistentDataTypes.itemStack,
                                )
                            }.toTypedArray()
                }

                val logEntries =
                    if (meta.persistentDataContainer.has<Array<PersistentDataContainer>, Array<PersistentDataContainer>>(
                            plugin.keys.logEntries,
                            PersistentDataType.TAG_CONTAINER_ARRAY,
                        )
                    ) {
                        meta.persistentDataContainer.get<Array<PersistentDataContainer>, Array<PersistentDataContainer>>(
                            plugin.keys.logEntries,
                            PersistentDataType.TAG_CONTAINER_ARRAY,
                        )
                            ?.map { compoundTag -> NuminousLogEntry.fromCompoundTag(plugin, compoundTag) }
                    } else {
                        listOf()
                    }
                return NuminousItemStack(itemType, amount, inventoryContents, logEntries)
            }

            @JvmStatic
            fun deserialize(serialized: Map<String, Any?>): NuminousItemStack {
                val itemService =
                    Services.INSTANCE.get(
                        NuminousItemService::class.java,
                    )
                val inventory = serialized["inventory-contents"] as List<ItemStack?>?
                val itemType =
                    itemService.getItemTypeById(serialized["item-type"] as String)
                        ?: throw IllegalStateException("Item type " + serialized["item-type"] + " does not exist")
                return NuminousItemStack(
                    itemType,
                    (serialized["amount"] as Number).toInt(),
                    inventory?.toTypedArray(),
                    serialized["log-entries"] as List<NuminousLogEntry>?,
                )
            }
        }
    }
